libqtile/widget/bluetooth_extra.py

```diff
@@ -0,0 +1,634 @@
+        _BluetoothBase.__init__(self, path, interface, properties_interface, widget)
+        self._connected = False
+        self._paired = False
+        self._status = 2  # 0: connected, 1: paired & unconnected, 2: unpaired
+        self._adapter = ""
+        self.has_name = False
+        self.battery_device = None
+        self._battery = 0
+
+    @catch_error("Unable to connect to device: {self.name}.")
+    async def connect(self):
+        await self.interface.call_connect()
+
+    @catch_error("Unable to disconnect device: {self.name}.")
+    async def disconnect(self):
+        await self.interface.call_disconnect()
+
+    @catch_error("Unable to pair device: {self.name}.")
+    async def pair_and_connect(self):
+        await self.interface.call_pair()
+        await self.connect()
+
+    async def action(self):
+        """Helper method to call appropriate method based on device status."""
+        if self.connected:
+            await self.disconnect()
+        elif self.paired and not self.connected:
+            await self.connect()
+        elif not self.paired:
+            await self.pair_and_connect()
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def connected(self):
+        return self._connected
+
+    @property
+    def paired(self):
+        return self._paired
+
+    @property
+    def status(self):
+        return self._status
+
+    @property
+    def battery(self):
+        if self.battery_device:
+            return self._battery
+        return ""
+
+    @property
+    def adapter(self):
+        a = self.widget.adapters.get(self._adapter)
+        if a:
+            return a.name
+        else:
+            return "Unknown"
+
+    def add_battery(self):
+        """Triggers adding battery interface."""
+
+        def refresh(_):
+            if self.battery_device:
+                create_task(self.update_props())
+
+        task = create_task(self.get_battery())
+        task.add_done_callback(refresh)
+
+    def remove_battery(self):
+        self.battery_device = None
+        self._battery = 0
+
+    async def get_battery(self):
+        proxy = await self.widget.get_proxy(self.path)
+        try:
+            self.battery_device = proxy.get_interface(BLUEZ_BATTERY)
+        except InterfaceNotFoundError:
+            return
+
+    async def update_props(self, setup=False):
+        """Refresh all the properties for the device."""
+        # Some devices don't report a name so we fall back to the device address
+        try:
+            self._name = await self.interface.get_name()
+            self.has_name = True
+        except (AttributeError, DBusError):
+            self._name = await self.interface.get_address()
+            self.has_name = False
+
+        self._connected = await self.interface.get_connected()
+        self._paired = await self.interface.get_paired()
+        self._adapter = await self.interface.get_adapter()
```

Comment:
Do these need to happen in series? If not, the three awaitables could be awaited simultaneously

Reply:
How would you propose doing that? I doubt there's any meaningful performance benefit of doing this.

Reply:
> How would you propose doing that?

Like this:
```py
self._connected, self._paired, self._adapter = asyncio.gather(
    self.interface.get_connected(),
    self.interface.get_paired(),
    self.interface.get_adapter(),
)
```

>  I doubt there's any meaningful performance benefit of doing this.

You're probably right :) unless for some reason your dbus connection was slow (is remote dbus a thing?)

Reply:
Well that looks neat. I'll make that change (although I suspect I need to `await` the `.gather` object.

And, yes, remote dbus is a thing but I can't imagine people would be using it much (and we don't provide the option of specifying a different dbus connection path).